File: Framework_Poison-Rat/rat.rb
require 'socket'
require 'openssl'

# Configurações de conexão
host = '127.0.0.1'
port = 443
timeout = 10

# Configuração para criptografia segura (TLS)
context = OpenSSL::SSL::SSLContext.new
context.verify_mode = OpenSSL::SSL::VERIFY_PEER

# Criando um socket seguro com TLS
socket_cliente = TCPSocket.new(host, port)
ssl_socket = OpenSSL::SSL::SSLSocket.new(socket_cliente, context)
ssl_socket.connect

# Configurações de criptografia AES-GCM
chave = 'sua-chave-secreta-de-32-bytes12345678'
iv_tamanho = 16 # AES GCM IV size

# Função para dividir, criptografar e enviar pacotes
def enviar_pacote_criptografado(socket, mensagem, chave, iv_tamanho)
  tamanho_pacote = 512
  pacotes = mensagem.scan(/.{1,#{tamanho_pacote}}/)

  pacotes.each_with_index do |pacote, contador|
    iv = OpenSSL::Random.random_bytes(iv_tamanho)
    pacote_com_nonce = "#{contador}|#{pacote}"

    # Criptografa com AES-256-GCM
    pacote_criptografado = OpenSSL::Cipher.new('aes-256-gcm').encrypt
    pacote_criptografado.key = chave
    pacote_criptografado.iv = iv
    pacote_final = pacote_criptografado.update(pacote_com_nonce) + pacote_criptografado.final

    # Anexa IV e Tag de autenticação
    socket.write("#{iv}#{pacote_final}")
  end
end

# Exemplo de mensagem a ser enviada
mensagem = "Olá, servidor! Esta é uma mensagem longa para testar o envio em pacotes."
enviar_pacote_criptografado(ssl_socket, mensagem, chave, iv_tamanho)

# Manter a comunicação aberta e ler resposta do servidor enquanto estiver ativo
while true
  resposta = ssl_socket.read(1024)
  break if resposta.nil? || resposta.empty?

  puts resposta
  sleep(0.1) # Aguarda 100ms antes de tentar novamente
end

# Fechar a conexão
ssl_socket.close

File: Framework_Poison-Rat/rat.py
import socket
import ssl
import os
import base64
import os.path
import struct
import time

# Configurações de conexão
host = '127.0.0.1'
port = 443
timeout = 10

# Criação de um contexto seguro com TLS
context = ssl.create_default_context()
context.check_hostname = True
context.verify_mode = ssl.CERT_REQUIRED
context.minimum_version = ssl.TLSVersion.TLSv1_2

# Criando um socket seguro com TLS
try:
    sock = socket.create_connection((host, port), timeout)
    secure_sock = context.wrap_socket(sock, server_hostname=host)
except Exception as e:
    print(f"Erro ao conectar: {e}")
    exit()

# Configurações de criptografia AES-GCM
chave = b'sua-chave-secreta-de-32-bytes12345678'  # 32 bytes para AES-256
iv_tamanho = 12  # IV tamanho para AES-GCM

def enviar_pacote_criptografado(socket, mensagem, chave, iv_tamanho):
    tamanho_pacote = 512
    pacotes = [mensagem[i:i + tamanho_pacote] for i in range(0, len(mensagem), tamanho_pacote)]

    for contador, pacote in enumerate(pacotes):
        iv = os.urandom(iv_tamanho)  # Gerar IV aleatório para cada pacote

        # Adiciona um contador ao pacote para evitar repetição
        pacote_com_nonce = f"{contador}|{pacote}".encode()

        # Criptografa com AES-256-GCM
        pacote_criptografado, tag = AES_GCM(chave).encrypt(iv, pacote_com_nonce)

        # Anexa IV e Tag de autenticação
        pacote_final = base64.b64encode(iv + tag + pacote_criptografado).decode()

        # Envia o pacote criptografado
        socket.sendall(pacote_final.encode())

# Função para manter a comunicação aberta e ler resposta do servidor enquanto estiver ativo
def manter_comunicacao(secure_sock):
    while True:
        data = secure_sock.recv(1024)
        if not data:
            break
        print(data.decode())

        time.sleep(0.1)  # Aguarda 100ms antes de tentar novamente

# Exemplo de mensagem a ser enviada
mensagem = "."
enviar_pacote_criptografado(secure_sock, mensagem, chave, iv_tamanho)

manter_comunicacao(secure_sock)

# Fechar a conexão
secure_sock.close()
